### PR TITLE
Replace unittest assertmethods with plain assert

### LINKED ISSUE #7185: Replace `unittest` assert methods with plain `assert`
## Details

Replace unittest's assert methods with plain `assert`.

## Example

```diff
class TestExample(unittest.TestCase):
    def test_foo(self):
-        self.assertEqual(1, 2)
+        assert 1 == 2
```

## Motivation

1. `assert` (rewritten by pytest) gives a better error message than unittest assert methods.
2. We'll migrate unittest test suites to pytest.
3. For code consistency.

## Instructions

1. Run `pip install libcst` for automatic code transformation.
2. Save the code below (which should transform all the assert methods in our code base) in `a.py`. Make sure that searching `self.assert[A-Z]` (regular expression) matches nothing. 
3. Run `python a.py $(git ls-files "tests/**/*.py")` in the repository root.
4. Run `black .`.

```python
import libcst as cst
import pathlib
import argparse
import difflib


class AssertMethodTransformer(cst.CSTTransformer):
    @staticmethod
    def is_unittest_assert_method(node: cst.Call, name: str) -> bool:
        return (
            isinstance(node.func, cst.Attribute)
            and isinstance(node.func.value, cst.Name)
            and node.func.value.value == "self"
            and isinstance(node.func.attr, cst.Name)
            and node.func.attr.value == name
        )

    def leave_Call(self, original_node: cst.Call, updated_node: cst.Call) -> cst.Assert:
        if (
            AssertMethodTransformer.is_unittest_assert_method(original_node, "assertEqual")
            or AssertMethodTransformer.is_unittest_assert_method(original_node, "assertListEqual")
            or AssertMethodTransformer.is_unittest_assert_method(
                original_node, "assertSequenceEqual"
            )
        ):
            left, right = original_node.args
            left = left.value
            right = right.value
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=left,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.Equal(),
                            comparator=right,
                        )
                    ],
                )
            )
            return assert_node

        if AssertMethodTransformer.is_unittest_assert_method(original_node, "assertCountEqual"):
            # `self.assertCountEqual({1, 2, 3}, x)` can converted to `assert {1, 2, 3} == x`
            if any(isinstance(a.value, cst.Set) for a in original_node.args):
                print(original_node.args[0].value.elements)
                left = original_node.args[0].value
                right = original_node.args[1].value
                assert_node = cst.Assert(
                    test=cst.Comparison(
                        left=left,
                        comparisons=[
                            cst.ComparisonTarget(
                                operator=cst.Equal(),
                                comparator=right,
                            )
                        ],
                    )
                )
                return assert_node
            # `self.assertCountEqual([], x)` or `self.assertCountEqual([1], x)` can be converted to
            # assert [] == x or assert [1] == x
            if any(
                isinstance(a.value, (cst.List, cst.Tuple)) and len(a.value.elements) < 2
                for a in original_node.args
            ):
                left = original_node.args[0].value
                right = original_node.args[1].value
                assert_node = cst.Assert(
                    test=cst.Comparison(
                        left=left,
                        comparisons=[
                            cst.ComparisonTarget(
                                operator=cst.Equal(),
                                comparator=right,
                            )
                        ],
                    )
                )
                return assert_node

            # `self.assertCountEqual([1, 2], x)` can't be converted to `assert [1, 2] == x`.
            # because x can be [2, 1]. We can convert it to either
            # `assert sorted([1, 2]) == sorted(x)` or `assert {1, 2} == set(x)`.
            # We choose the former because the latter allows x to contain duplicates.
            # For example, if x is the return value of `search_experiments`, it should not contain duplicates.
            left, right = original_node.args
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=cst.Call(
                        func=cst.Name("sorted"),
                        args=[left],
                    ),
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.Equal(),
                            comparator=cst.Call(
                                func=cst.Name("sorted"),
                                args=[left],
                            ),
                        )
                    ],
                )
            )
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertNotEqual"):
            left, right = original_node.args
            left = left.value
            right = right.value
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=left,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.NotEqual(),
                            comparator=right,
                        )
                    ],
                )
            )
            return assert_node
        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertIsInstance"):
            left, right = original_node.args
            assert_node = cst.Assert(test=cst.Call(func=cst.Name("isinstance"), args=[left, right]))
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertTrue"):
            assert_node = cst.Assert(test=original_node.args[0].value)
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertIsNone"):
            (arg,) = original_node.args
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=arg.value,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.Is(),
                            comparator=cst.Name("None"),
                        )
                    ],
                )
            )
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertIsNotNone"):
            (arg,) = original_node.args
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=arg.value,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.IsNot(),
                            comparator=cst.Name("None"),
                        )
                    ],
                )
            )
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertIn"):
            left, right = original_node.args
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=left.value,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.In(),
                            comparator=right.value,
                        )
                    ],
                )
            )
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertLessEqual"):
            left, right = original_node.args
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=left.value,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.LessThanEqual(),
                            comparator=right.value,
                        )
                    ],
                )
            )
            return assert_node

        elif AssertMethodTransformer.is_unittest_assert_method(original_node, "assertAlmostEqual"):
            left, right = original_node.args[:2]
            assert_node = cst.Assert(
                test=cst.Comparison(
                    left=left.value,
                    comparisons=[
                        cst.ComparisonTarget(
                            operator=cst.Equal(),
                            comparator=cst.Call(
                                cst.Attribute(value=cst.Name("pytest"), attr=cst.Name("approx")),
                                args=[right],
                            ),
                        )
                    ],
                )
            )
            return assert_node

        return original_node


def transform_file(path: pathlib.Path) -> None:
    src = path.read_text()
    source_tree = cst.parse_module(src)
    modified_tree = source_tree.visit(AssertMethodTransformer())
    if not modified_tree.deep_equals(source_tree):
        print("".join(difflib.unified_diff(src.splitlines(1), modified_tree.code.splitlines(1))))
        path.write_text(modified_tree.code)
        print(f"Transformed {path}")


if __name__ == "__main__":
    # python a.py $(git ls-files "tests/**/*.py")
    parser = argparse.ArgumentParser(description="Process some integers.")
    parser.add_argument("files", metavar="N", nargs="+", help="an integer for the accumulator")
    args = parser.parse_args()
    for f in args.files:
        print("Processing", f)
        transform_file(pathlib.Path(f))
```

## Out of scope:

- Swapping actual and expected values.

### PR DESCRIPTION
**Pull request by ayushthe1**
Signed-off-by: ayush_gitk <ayushsharmaa101@gmail.com>

## Related Issues/PRs
fixes #7185 

## What changes are proposed in this pull request?

All unittest `assertmethods` are replaced with plain `assert`

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
@ayushthe1 Thanks for the PR! I found an issue in the transformation script and fixed it. Can you reset the changes and run this command?

```
pip install libcst && python -c "$(curl https://gist.githubusercontent.com/harupy/7fdf544b78d62720ee67473b1b5a6576/raw/ca4bd77d502b730ce1877704cf3ecae9ec1c0078/transform_assert_methods.py)" $(git ls-files "tests/**/*.py") && black .
```

File: tests/pipelines/cards/test_histogram_generator.py [hunk 1]
```diff
@@ -19,14 +19,20 @@ def assert_histogram(
...
-        self.assertEqual(len(expected.buckets), len(actual.buckets))
+        assert expected.type == actual.type
+        assert len(expected.buckets) == len(actual.buckets)
         for i in range(len(expected.buckets)):
             actual_bucket = actual.buckets[i]
             expected_bucket = expected.buckets[i]
-            self.assertAlmostEqual(actual_bucket.low_value, expected_bucket.low_value, 2)
-            self.assertAlmostEqual(actual_bucket.high_value, expected_bucket.high_value, 2)
-            self.assertAlmostEqual(actual_bucket.sample_count, expected_bucket.sample_count, 2)
+            assert actual_bucket.low_value == pytest.approx(
```
Comment (harupy) [inline tests/pipelines/cards/test_histogram_generator.py:27]: Can you add `import pytest`?

**Comment by ayushthe1** [comment]
Thanks @harupy .I commited the changes as you pointed out.

**Comment by harupy** [comment]
@ayushthe1 Can you fix the conflict?

**Comment by ayushthe1** [comment]
@harupy ,i fixed the conflict .Do tell if something more needs to be done.

**Review by harupy — APPROVED** [review]

**Comment by harupy** [comment]
@ayushthe1 Can we adjust the tolerance of `pytest.approx`?

https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-approx

**Comment by ayushthe1** [comment]
Run black .

**Comment by ayushthe1** [comment]
hey @harupy ,i adjusted the `rel` parameter value to `1e-3` in [test_histogram_generator.py](https://github.com/mlflow/mlflow/blob/a951d7cb5f73f4757c694e2bb639419fea2d565f/tests/recipes/cards/test_histogram_generator.py#L28) to change the tolerance of `pytest.approx` .
Also ,DCO tests were failing and to resolve it ,it instructed me to rebase my branch causing so much clutter of commits above.

File: tests/recipes/cards/test_histogram_generator.py [hunk 2]
```diff
@@ -19,14 +20,16 @@ def assert_histogram(
...
             actual_bucket = actual.buckets[i]
             expected_bucket = expected.buckets[i]
-            self.assertAlmostEqual(actual_bucket.low_value, expected_bucket.low_value, 2)
-            self.assertAlmostEqual(actual_bucket.high_value, expected_bucket.high_value, 2)
-            self.assertAlmostEqual(actual_bucket.sample_count, expected_bucket.sample_count, 2)
+            assert actual_bucket.low_value == pytest.approx(expected_bucket.low_value, rel=1e-3)
+            assert actual_bucket.high_value == pytest.approx(expected_bucket.high_value, rel=1e-3)
+            assert actual_bucket.sample_count == pytest.approx(
+                expected_bucket.sample_count, rel=1e-3
+            )
```
Comment (harupy) [inline tests/recipes/cards/test_histogram_generator.py:32]:
```suggestion
            assert actual_bucket.low_value == pytest.approx(expected_bucket.low_value, rel=1e-2)
            assert actual_bucket.high_value == pytest.approx(expected_bucket.high_value, rel=1e-2)
            assert actual_bucket.sample_count == pytest.approx(
                expected_bucket.sample_count, rel=1e-2
            )
```

File: tests/recipes/cards/test_histogram_generator.py [hunk 3]
```diff
@@ -39,18 +42,21 @@ def test_invalid_quantiles(self):
...
+            is None
         )
         # When (len(quantiles) - 1) % num_buckets != 0, quantiles is considered as invalid.
-        self.assertIsNone(
+        assert (
             histogram_generator.generate_equal_height_histogram(
                 quantiles=[1, 2, 3, 4, 5], num_buckets=3
             )
+            is None
+
```
Comment (harupy) [inline tests/recipes/cards/test_histogram_generator.py:59]:
```suggestion
```

**Comment by harupy** [comment]
@ayushthe1 I fixed the `pytest.approx` error and cleaned up the commit history.

**Pull request merged**